### DISCUSSION TITLE
css는 어떤 방식으로 사용하는 것이 효율적일까요?

### DISCUSSION THREAD

**Discussion by taetaeil** [seongpil0948/framework-next#21] [Ideas]
_현재(11.06) 결정된 것은 없으며 다양한 레퍼런스를 찾아보며 방법을 정리 중 입니다._

1. NEXT JS 환경에서 스타일을 적용 방법론
  - CLSX 플러그인 &#9734; &#9734; &#9734;
   -  클래스 이름을 동적으로 생성하고 관리하는 데 도움을 주는 라이브러리로, 여러 조건에 따라 다른 클래스 이름을 추가하거나 제거할 수 있도록 도와줍니다. (vue의 클래스 바인딩과 역할이 같습니다.)
  - Tailwind merge
   - tailwind를 사용할 경우 불러오는 컴포넌트에 적용되어 있는 스타일 중 일부를 변경하고 싶을 경우 css의 캐스케이딩 방식의 한계를 극복하고 원하는 클래스를 넣을 수 있는 기능(override)
  - tailwind varient
  - CSS module

2. 효율적인 CSS 사용 방안
  - Tailwind의 단독사용
  - SCSS의 단독사용
  - Tailwind와 scss의 혼합 사용

**Comment by seongpil0948** [comment]
여러 상황들을 가정하여, 프로젝트에 맞게 퍼블리싱을 진행 할 수 있도록
고려해보았습니다.
계속 논의하시죠

## 1. 디자인에 맞게 속성 및 컬러 변경
```ts
// tailwind.config.js
const {nextui} = require("@nextui-org/react");

/** @type {import('tailwindcss').Config} */
module.exports = {
  plugins: [
    nextui({
      layout: {
        disabledOpacity: "0.3", // opacity-[0.3]
        radius: {
          small: "2px", // rounded-small
          medium: "4px", // rounded-medium
          large: "6px", // rounded-large
        },
        borderWidth: {
          small: "1px", // border-small
          medium: "1px", // border-medium
          large: "2px", // border-large
        },
      },
      themes: {
        light: {
          colors: {
            primary: {
              DEFAULT: "#BEF264",
              foreground: "#000000",
            },
            focus: "#BEF264",
        },
        dark: {},
      },
    }),
  ],
};
```

## 2. 테마 값을 모듈로 사용 할 경우
```ts
import {commonColors, semanticColors} from "@nextui-org/theme";

console.log(commonColors.white); // #FFFFFF
console.log(commonColors.blue[500]); // #006FEE

console.log(semanticColors.dark.warning.DEFAULT); // #FFC107
console.log(semanticColors.light.primary.DEFAULT); // #006FEE
```

## 3. 타 클래스 선언 혹은 CSS 변수로 사용 할 경우
```ts
/* With default prefix */
.my-component {
  background-color: var(--nextui-primary-500);
}
/*  With custom prefix */
.my-component {
  background-color: var(--myapp-primary-500);
}
```
## 4. 아예 컴포넌트를 갈아엎을 경우
https://nextui.org/docs/customization/custom-variants
링크 참고


## 참고 자료
- https://tailwindcss.com/docs/configuration
- https://nextui.org/docs/customization/theme

**Reply by taetaeil** [reply]
next ui 커스텀 방법은 1번과 4번을 활용하는 것이  좋은 방법이라고 생각됩니다.
